### DISCUSSION TITLE
stage 1: kedro parameters <-> dvc parameters

### DISCUSSION THREAD

**Discussion by shaunc** [FactFiber/kedro-dvc#12] [Ideas]
## Representation

[Kedro parameters][k-parameters] are configured project `config/base`/`config/local` and are located relative to there or to project root (searched in order `local`, `base`,  then project root). Parameters are stored in yaml files.

Kedro allows parameters to be consumed directly in python code. Or they can be [consumed as node inputs][k-param-use]. Credentials are managed separately, and can be [used in the data catalog][k-cred-catalog]. Secure credentials are intended to be placed in `config/local/credentials` (`config/local` should be in `.gitignore`).

Kedro parameters are by default inside the configuration environment (typically in `conf/base` or `conf/local`). They are referenced in inputs using the syntax `params:<key>` to reference individual parameters. (Unlike DVC, it would seem that the key is not a dottedname, but references a top-level entity in parameters.

DVC supports [a wide range of formats][d-parameters] for parameter specification. They are referenced in pipeline files. Credentials are not tracked, as reading and writing of files is not modelled by dvc.

DVC specifies parameters in two different manners -- in [`params` section in each stage][d-parameters], and via templates [`vars` section][d-param-vars], which can be ether global to pipeline, or local to stage.

DVC has a default `params` file (`params.yaml`, but other formats are also supported). The stage `params` section is a list of parameters, keys may be specified directly, or they may be specified as a dictionary entry:

```yaml
stage:
  params:
  - <key>
  - <parameter file name>:
    - <key>
```

In either case, `<key>` is a dotted name referencing a possibly nested value in the parameter file.

In `dvc.yaml` templates, parameters can be used not only as inputs to stage execution, but also can vary other stage metadata.

## Implementation

Only parameters used as node inputs can be tied to experiments. Kedro-DVC will include parameters referenced by nodes using the `params:<key>` syntax in `dvc.yaml` stage definitions.

[k-parameters]: https://kedro.readthedocs.io/en/latest/04_kedro_project_setup/02_configuration.html#parameters
[k-param-use]: https://kedro.readthedocs.io/en/latest/04_kedro_project_setup/02_configuration.html#use-parameters
[k-cred-catalog]: https://kedro.readthedocs.io/en/latest/05_data/01_data_catalog.html#feeding-in-credentials

[d-parameters]: https://dvc.org/doc/start/metrics-parameters-plots#defining-stage-parameters
[d-param-vars]: https://dvc.org/doc/user-guide/project-structure/pipelines-files#templating

Prev | Up | Next
--- | --- | ---
[data][disc-data] | [stage 1][disc-stage-1] | [artifacts][disc-art]

[disc-stage-1]: https://github.com/FactFiber/kedro-dvc/discussions/6
[disc-data]: https://github.com/FactFiber/kedro-dvc/discussions/9
[disc-param]: https://github.com/FactFiber/kedro-dvc/discussions/12
[disc-pipeline]: https://github.com/FactFiber/kedro-dvc/discussions/11
[disc-art]: https://github.com/FactFiber/kedro-dvc/discussions/13
[disc-exp]: https://github.com/FactFiber/kedro-dvc/discussions/14

**Comment by antonymilne** [comment]
Just a few things to add here (some I already mentioned on discord, but let me put them here just for completeness).

1. Under the hood, kedro's [`ConfigLoader`](https://kedro.readthedocs.io/en/stable/kedro.config.ConfigLoader.html) uses [anyconfig](https://pypi.org/project/anyconfig/). This actually supports a whole range of of configuration formats (listed in our documentation is yaml, json, ini, pickle, xml, properties, but maybe anyconfig does even more now). With that said, yaml is overwhelmingly the most popular option (no exaggeration when I say that I'd guess that >99.99% of kedro config files are yaml)
2. Referencing subkeys of parameters is possible and looks quite similar to DVC: `params:key.subkey.subsubkey`
3. "Kedro allows parameters to be consumed directly in python code. Or they can be [consumed as node inputs](https://kedro.readthedocs.io/en/latest/04_kedro_project_setup/02_configuration.html#use-parameters)." While there's nothing technically stopping you from manually loading up a parameters file and consuming it in your node code, this is _very strongly_ discouraged and shouldn't ever really happen. So "Only parameters used as node inputs can be tied to experiment" is a very reasonable limitation
4. `ConfigLoader` is something which [does have a well-defined interface](https://github.com/kedro-org/kedro/blob/develop/kedro/config/abstract_config.py), and kedro ships with [two implementations of that](https://kedro.readthedocs.io/en/stable/kedro.config.html). `TemplatedConfigLoader` enables templating through injection of global variables and Jinja2
5. `ConfigLoader` is probably the easiest and most common kedro component to swap out in settings.py, e.g. people sometimes want to change the run environment hierarchy from `local > base` to something more complex like `local > env-subenv > env > base`. Not sure if a custom config loader will be any use on kedro-dvc, but just to say that if it is required then it should be relatively easy.

In general the topic of configuration (usually in relation to catalog entries rather than parameters) is quite a hot topic in kedro and there's lots of different opinions on it, e.g. https://github.com/kedro-org/kedro/issues/1076, https://github.com/kedro-org/kedro/issues/891, https://github.com/kedro-org/kedro/issues/770, https://github.com/kedro-org/kedro/issues/1011. So chances are if you need to do something a bit out of the ordinary here, someone has considered how to do it already at some point.

**Reply by shaunc** [reply]
1. Cool! (You might want to put in a few more mentions in the doc of `anyconfig`....)
2. Also cool! Yes -- you mentioned on discord. I had missed that.
3. What I meant by "consumed directly in python code" is that you can use config loader completely outside of a pipeline -- but `kedro-dvc` won't know or care about that. I meant by "node inputs" the regular pipeline-mediated inputs to nodes, not getting parameters directly, bypassing the pipeline.

vs 4 and 5 as well as general comment -- especially as this is a hot topic I'm hoping to avoid imposing extra requirements. I need to find out what parameters are used so I can tell dvc about them in the `dvc.yaml`.  But I want to do that without restricting project users if possible.

Perhaps the best idea would be to just try to "run" the pipeline, then look at context in the `before_pipeline_run` hook before aborting to write the `dvc.yaml` file for the pipeline ... hmm... @AntonyMilneQB -- is there enough information available to `before_pipeline_run` to figure out source files for all the parameters?

**Reply by antonymilne** [reply]
Unfortunately I don't think there is. The only thing that's present in `catalog.datasets` are the names of the datasets, which will include all the possible `params:*` patterns that would be valid node inputs. The information about which files these were originally defined in has long gone by that point. In fact this information is never really available within kedro. [This is how parameters are loaded](https://github.com/kedro-org/kedro/blob/develop/kedro/framework/context/context.py#L244-L246) in the first place. If you follow that does does through to lower level details then ultimately the best you can easily achieve is to get a list of _all_ parameters file that kedro looks at, as we do [here](https://github.com/kedro-org/kedro/blob/74301c70c31eef71dd372fd3b13afb91b6f297d3/kedro/config/common.py#L241). There's never really a mapping of which parameters are defined in which files available anywhere.

**Reply by shaunc** [reply]
Hmm... and then [`_path_lookup`](https://github.com/kedro-org/kedro/blob/74301c70c31eef71dd372fd3b13afb91b6f297d3/kedro/config/common.py#L241) is called by [`_lookup_config_filepaths`](https://github.com/kedro-org/kedro/blob/74301c70c31eef71dd372fd3b13afb91b6f297d3/kedro/config/common.py#L189) which sorts the (arbitrarily ordered) files, and that order determines how parameters are resolved. Then that in turn is called by [`_get_config_from_patterns`](https://github.com/kedro-org/kedro/blob/74301c70c31eef71dd372fd3b13afb91b6f297d3/kedro/config/common.py#L53) which throws out the filepaths. This is finally called by `ConfigLoader` and/or `TemplatedConfigLoader`. 

Sigh... so it would seem that we'll have to create this logic over again. 

Any chance you'd accept a PR to bubble up a map from key -> filepath and use it in specializations of `AbstractConfigLoader`? (An instance of which would be available in the context?)

Aside -- it would seem that would be nice if [this](https://github.com/kedro-org/kedro/blob/74301c70c31eef71dd372fd3b13afb91b6f297d3/kedro/config/config.py#L98) were memoizing? (Or isn't it called very much?)
```python
    def get(self, *patterns: str) -> Dict[str, Any]:
        return _get_config_from_patterns(
            conf_paths=self.conf_paths, patterns=list(patterns)
        )
```

**Reply by antonymilne** [reply]
Very reasonable though that sounds, to be honest I think such a PR is unlikely to be accepted, because the whole topic of config in kedro is such a big one that we're trying to deal with it holistically and not change just small pieces of it any more. Sorry! Note that even if the key -> filepath map were available after `ConfigLoader` you would still need to make it available in `catalog` somehow. Unfortunately I think you'll have to come up with some workaround here:
* recreate the logic yourself
* or write a new `ConfigLoader` (and also `DataCatalog`) which do what you need them to
* link the dependency of a parameter input to _all_  available parameter files rather than just the source one

**Reply by antonymilne** [reply]
N.B. even though the last suggestion sounds very crude, it's worth noting that even if you did have the key -> filepath map available this still isn't ideal. e.g. if my node input is `params:key.subkey` then what I really care about is whether the hash of that particular value has changed, not the hash of the file it's contained in, which could contain any number of irrelevant parameters that I'm changing. A key -> filepath map would only be sufficient if you break your parameters files up so that each node's parameter dependencies is stored in a separate file (which might not be a bad way to work but definitely isn't always the case in kedro).

**Reply by shaunc** [reply]
> A key -> filepath map would only be sufficient if you break your parameters files up...

My thought here was to use "file" dependencies for dvc, then (in stage 2, perhaps) do fine-grained skip based on particular parameter. (But how would I know what the old value is? Am I caching them somewhere?)

**HOWEVER** -- I just may have hit on a way to avoid this who morass -- tell me what you think!

What if I just write new parameter files in `conf/kedro_dvc/` and point dvc at them? When run starts, before I ask DVC which nodes to run, I can write in the most recent parameters.

I can have one file per parameter. (Perhaps this could be configurable someday....)
